src/passes/OnceReduction.cpp

```diff
@@ -0,0 +1,414 @@
+  // global if so, or an empty name otherwise.
+  //
+  // TODO: If the "once" function is inlined, this pattern can show up in random
+  //       places, and we can look for it there as well.
+  Name getOnceGlobal(Expression* body) {
+    // Look the pattern mentioned above:
+    //
+    //  function foo() {
+    //    if (foo$once) return;
+    //    foo$once = 1;
+    //    ...
+    //
+    auto* block = body->dynCast<Block>();
+    if (!block) {
+      return Name();
+    }
+    auto& list = block->list;
+    if (list.size() < 2) {
+      return Name();
+    }
+    auto* iff = list[0]->dynCast<If>();
+    if (!iff) {
+      return Name();
+    }
+    auto* get = iff->condition->dynCast<GlobalGet>();
+    if (!get) {
+      return Name();
+    }
+    if (!iff->ifTrue->is<Return>() || iff->ifFalse) {
+      return Name();
+    }
+    auto* set = list[1]->dynCast<GlobalSet>();
+
+    // Note that we have already checked the set's value earlier, but we do need
+    // it to not be unreachable (so it is actually set).
+    if (!set || set->name != get->name || set->type == Type::unreachable) {
+      return Name();
+    }
+    return get->name;
+  }
+
+private:
+  OptInfo& optInfo;
+};
+
+// Information in a basic block. We track relevant expressions, which are calls
+// calls to "once" functions, and writes to "once" globals.
+struct BlockInfo {
+  std::vector<Expression*> exprs;
+};
+
+struct Optimizer
+  : public WalkerPass<CFGWalker<Optimizer, Visitor<Optimizer>, BlockInfo>> {
+  bool isFunctionParallel() override { return true; }
+
+  Optimizer(OptInfo& optInfo) : optInfo(optInfo) {}
+
+  Optimizer* create() override { return new Optimizer(optInfo); }
+
+  void visitGlobalSet(GlobalSet* curr) {
+    if (currBasicBlock) {
+      currBasicBlock->contents.exprs.push_back(curr);
+    }
+  }
+
+  void visitCall(Call* curr) {
+    if (currBasicBlock) {
+      currBasicBlock->contents.exprs.push_back(curr);
+    }
+  }
+
+  void doWalkFunction(Function* func) {
+    using Parent =
+      WalkerPass<CFGWalker<Optimizer, Visitor<Optimizer>, BlockInfo>>;
+
+    // Walk the function to builds the CFG.
+    Parent::doWalkFunction(func);
+
+    // Build a dominator tree, which then tells us what to remove: if a call
+    // appears in block A, then we do not need to make any calls in any blocks
+    // dominated by A.
+    DomTree<Parent::BasicBlock> domTree(basicBlocks);
+
+    // Perform the work by going through the blocks in reverse postorder and
+    // filling out which "once" globals have been written to.
+    auto numBlocks = basicBlocks.size();
+    if (numBlocks == 0) {
+      return;
+    }
+
+    // Each index in this vector is the set of "once" globals written to in the
+    // basic block with the same index.
+    std::vector<std::unordered_set<Name>> onceGlobalsWrittenVec;
+    onceGlobalsWrittenVec.resize(numBlocks);
+
+    for (Index i = 0; i < basicBlocks.size(); i++) {
```

Comment:
In case we don't switch the above to `basicBlocks.empty()`
```suggestion
    for (Index i = 0; i < numBlocks; i++) {
```